### ISSUE TITLE
Evaluate secret management options for this repository

### ISSUE THREAD

**Issue by yuvipanda** [berkeley-dsep-infra/datahub#596]
We strive to keep our deployments code as public as possible for the
following reasons:

1. GSIs, TAs and professors can self-serve environment and config changes
   with pull requests in a workflow familiar to anyone who has used
   GitHub in the past. Also makes it simple for them to see what is
   actually installed without having to bother IT staff.
2. Our deployment is pretty cutting edge, so other organizations and
   institutes can use it as a model of how to operate this.
3. Continuous Deployment providers like CircleCI are much easier to work
   with public repositories.

Having a public repository also forces us to take secrets management very
seriously. In our configuration, we need to keep several secrets:

1. Authentication API Keys (for Googel Auth, EdX, etc)
2. Cookie Secrets
3. Authentication tokens for talking between hub components
4. Student id lists for extra RAM (this is gonna go away soon)

Note that we don't have any HTTPS related secrets, since that is handled
automatically for us by [Let's Encrypt](https://letsencrypt.org/).

We wanna keep these secrets in the same git repository for ease of
deployment and management, but we wanna restrict access to these secrets.
Only our continuous deployment environment and IT personnel should have
access to these secrets.

This document is a formal evaluation of the various options we are considering.

### Unified with rest of configuration

Our configuration lives in this repository as YAML files, and so should our secrets.
Secrets are ultimately just configuration, and shouldn't be treated differently. The
JupyterHub chart doesn't treat them differently for this reason, so we should follow its
lead and not use a separate GUI application or what not for it.

### Access Control

We should be able to manage who can encrypt and decrypt the secrets in a consistent
fashion. There should be, at all times, a centralized list of people we know
to have access to this. The continuous deployment environment (currently CircleCI)
should be able to decrypt, but not encrypt our secretes.

### Revocation 

When folks leave, we should be able to revoke their access to secrets. Once revoked, you shouldn't be able to decrypt *historical* versions of the secrets either. This also means that if a *secret key* leaks, we should be able to protect new and existing secret decryption processes by doing a key revocation. It should also let us revoke keys easily.

### Prevent accidental leakage

Accidental leakage to plaintext on Git is a strong issue that needs good safeguards.
We must make sure that whatever solution we pick is designed to make that as hard
as possible, and fails safely - so that we might accidently encrypt *more* things than
we thought, not less.

### Easy to install and use

Setting up local decryption should be fairly painless, well documented and easy to
do across multiple platforms. If it is too cumbersome to use, people will not use it.

### git-crypt

[git-crypt](https://github.com/AGWA/git-crypt) provides 'transparent' encryption and
decryption of files in a git repository. You can specify paths that must always be
encrypted, and git-crypt will encrypt them during `git commit` and decrypt them during
`git checkout`. 

#### Unified with rest of configuration

In the deployment, config files are under `deployments/<name>/config` and secrets are
under `deployments/<name>/secrets`. Secrets can be YAML files, in which case they will
be seamlessly merged into config files during deployment. We can also encrypt arbitrary
other files there if needed.

#### Access Control

The most popular method of using git-crypt is with a single, symmetric key that is used
for encryption and decryption. Anyone who has access to this key can encrypt and decrypt
all secrets. This key needs to be shared very securely, using out of band methods like
keybase or ssh-vault or something a bit more hacky. If this key leaks, all prior data is
fully decryptable. There is also no way to know the list of people who have the symmetric
key at any given time.

git-crypt also has a GPG mode that fixes most of these issues, as long as participants
are able to set up and run GPG properly. List of keys that can be used to decrypt / encrypt
is present in the repository itself, and no key transfer is needed. 

#### Revocation

In the symmetric encryption mode, there is no way to revoke the key - there is only one shared key. Once the key leaks, *all* historical secrets can be decrypted - there is no way around that. Even key rotation isn't implemented - https://github.com/AGWA/git-crypt/issues/61. The security of the secret key is of extreme importance here. This is at odds with the fact we need to distribute it somewhat widely if we wanna have people and continuous deployment environments have access to this.

#### Accidental Leakage

git-crypt integrates strongly with git to prevent this. Directories to be encrypted are
specified in `.gitattributes`. They are encrypted on commit and decrypted on checkout,
if you have the key. This makes it very hard to accidentally commit plaintext files, as
long as you have .gitattributes set up properly in the repository. 

#### Easy to install and use

git-crypt is available in Ubuntu repositories since 18.10, and there is a homebrew
formula for macOS. However, there are no official builds for Windows, nor is it clear
if it can build there.

To use GPG with git-crypt requires a fully functional GPG setup locally, as well as
proper GPG key security for all participants involved.

### sops

[sops](https://github.com/mozilla/sops) is a project from Mozilla for encrypting secrets 
for deployment using key management systems found in modern cloud platforms or GPG.

#### Unified with rest of configuration

In the deployment, config files are under `deployments/<name>/config` and secrets are
under `deployments/<name>/secrets`. Secrets can be YAML files, in which case they will
be seamlessly merged into config files during deployment. We can also encrypt arbitrary
other files there if needed.

YAML files are [treated differently](https://github.com/mozilla/sops#important-information-on-types),
so only values are encrypted. This is an advantage most of the time, but does require us to
be careful to not leak any information in the keys. We can override this special
behavior if needed.

#### Access Control

sops is commonly used alongside a Key Management System found in a cloud system,
such as [GCP](https://cloud.google.com/kms/), [AWS](https://aws.amazon.com/kms/) or [Azure](https://azure.microsoft.com/en-in/services/key-vault/). Access
control is thus handled at the cloud platform level. These access control systems let us provide fine grained, centralized permissions on who can access these secrets when and how. We can also revoke access to folks when needed.

Since the encryption info for each file is self contained, you can use different keys to encrypt different files - and give people fine-grained access to files.

The GPG mode of sops is similar to the GPG mode of git-crypt.

#### Revocation

Access controls are centrally managed, and when revoked they'll prevent decryption of historical and current secrets. This is because the encryption and decryption actually happen in the cloud - the secret key is never on the user's computer. This is pretty great, since we can revoke / rotate secret keys easily.

#### Accidental leakage

Sops doesn't really integrate with git - it can show pretty diffs but that's it. Creating and editing secrets thus requires running the 'sops' command manually. So to create or edit a secret file, you would have to run `sops <path to file>`. This opens your default editor (which you can set with EDITOR command), and then you can edit the file, and close it to save. So the files are encrypted *on your disk* as well, and only decrypted explicitly when you need to edit them.

This also means we'll have to integrate sops support into hubploy (our deployment utility) if we use sops. The secrets will have to be decrypted on the fly before deployment. This shouldn't be hard.

The downside is that we can't mark an entire directory as 'encrypted'. So if a user manually creates a file inside the 'secrets' directory and expects it to be encrypted, they are in for a surprise! 

*experimental evaluation with multiple people needed*

#### Easy to install and use

There are official builds for Windows, Linux and macOS. You need to explicitly call it from the commandline, which makes it less convenient to use than git-crypt. The `EDITOR` environment variable can be set to have it use any editor. The explicit action required to edit or view the secrets makes it clear that it is encrypted, but does require you to use the commandline to explicitly do it.

*Experimental evaluation with other people needed*

**Comment by yuvipanda** [comment]
/cc @felder and @ryanlovett. 

I decided to do a somewhat formal evaluation before figuring which secret management system to use going forward, given that we're expanding beyond the 2-3 people where our current system works. Let me know if you think we should involve other criteria, or involve other systems.

I've filled out evals for git-crypt since we've used it successfully here and in other projects. I'll play with sops over the coming days and fill in results here. Hopefully we'll be able to decide on a solution and switch to it soon - maybe even this week.

**Comment by yuvipanda** [comment]
I tested sop's git integration a little. It's not super well integrated - it is designed for the user to fully know that sops is being used while using it. With a config file present indicating which files are encrypted with which keys, you can create or edit a secret file with `sops <filename>`. This decrypts the file, lets you edit it, and re-crypts it in the editor of your choice. 

Initial thoughts:

1. You can encrypt different files with different keys from different projects. So you can encrypt the data8x secrets with a key from the data8x project, accessible only to people who are admins there. 
2. You have to explicitly use a separate command to see the secrets. I am not sure yet if this is a positive or negative. With git-crypt, it is just another file  that you can edit as usual.
3. It's file based rather than directory based. If the user adds a *new* file to a secrets/ directory *expecting* it to be kept secret, they are in for a surprise! Possible solutions here are a pre-commit hook checking for this, and a quick-fail CI that also checks for this - although at CI it is already too late to prevent that from leaking.

You can see how an encrypted file looks here: https://github.com/yuvipanda/sops-key-test/blob/master/data8x/secrets/test.yaml. You can see the various encrypted keys, but not values in the YAML. The file is also fully self-contained as it includes information on what key was used to encrypt it.

**Comment by yuvipanda** [comment]
sops encrypted YAML files also show the keys unencrypted, and the values are encrypted. So it looks like:

```
hub:
    jupyterhub:
        proxy:
            secretToken: ENC[AES256_GCM,data:azhcfKu9Ow==,iv:J/xnjpJmDprwQqf+kOk0ZVT00Hz5sVITEbsCRYRlr0c=,tag:xiuY5SrMPQgqncp52z9BWA==,type:str]
```

So the value of secretToken looks encrypted, and it is. However, there's nothing preventing people from just adding things there manually, in cleartext! They can then push this and nothing is stopping them.

**Comment by yuvipanda** [comment]
I did a quick check on what kinda commits touch the secret files for the various deployments.

1. datahub - 22 commits. 9 commits are memory bumps (fixed with #476). 5 are changing Google Auth secrets. 2 are secret rotations. 5 are mass refactors when we moved lots of things around to switch deployment methods. 1 is for adding a new secret for google sheets.

2. data8x - 5 commits. 3 are secrets for the gofer grading service. 2 are for initial setup. 1 is a test.

3. prob140 - 3 commits. 1 is initial setup, 1 is part of big refactor, and 1 is a Google Auth credential change.

4. math124 - 3 commits. 2 for initial setup, 1 is part of big refactor.

5. external - 2 commits, all for initial setup.

Almost all of this were done by me and Ryan, with one test commit from Jon Felder.

**Comment by yuvipanda** [comment]
Once #476 is done, the only things that should be in secret files are:

1. Authentication Credentials
2. Cookie Secret salts
3. Proxy Secret Token

Nothing else should really be in secrets. This means that we don't expect that many changes to these files, nor that many people will need to change them.

**Comment by yuvipanda** [comment]
@felder @ryanlovett I'm hoping both of you would be able to test out sops to see how usable it is. I've set up a test repo: https://github.com/yuvipanda/sops-key-test/blob/master/data8x/secrets/test.yaml

You should be able to:

1. Download sops for your platform: https://github.com/mozilla/sops/releases
2. Auth with `gcloud` to an account that has access to the data8x-scratch project.
3. Try decrypt / edit the file with `sops <filename>`, and see if you can edit it / view it, change it. On save, it should be encrypted on disk.
4. Try create a new file that is encrypted, with `sops <filename>` inside data8x/secrets. It automatically figures out which key to use based on .sops.yaml.

I *think* I've done all the evaluation I wanna do, and would love to hear your opinions of this after testing this out.

**Comment by ryanlovett** [comment]
@yuvipanda I should be able to get to this tomorrow.

**Comment by ryanlovett** [comment]
I was initially unable to decrypt the file, but after unsetting GOOGLE_APPLICATION_CREDENTIALS and doing `google auth application-default login`, it worked.

```
$ gcloud kms keys list --location global --keyring data8x-sops-test
NAME                                                                                                PURPOSE          LABELS  PRIMARY_ID  PRIMARY_STATE
projects/data8x-scratch/locations/global/keyRings/data8x-sops-test/cryptoKeys/data8x-sops-test-key  ENCRYPT_DECRYPT          1           ENABLED
```

Confirmed that `sops data8x/secrets/test.yaml` dropped me into an editor where I was able to change an unencrypted string. I catted the same file and the token value was encrypted.

I was able to create a new secret file in secrets/. So "If the user adds a new file to a secrets/ directory expecting it to be kept secret, they are in for a surprise!" is not true via your .sops.yaml glob as you indicated later. I tried using sops to create a new file outside of secrets/ but it properly errored.

Moving the authorization to the cloud is handy since you can grant access via the cloud portal rather than by exchanging a git-crypt secret file out of band. Since sops appears to handle the base functionality we need, I suppose the project management of secret access is the most important feature. This is critical if we needed many people to read/write secrets, but only moderately handy since we don't.

> However, there's nothing preventing people from just adding things there manually, in cleartext!

I vi'ed a secret file and inserted a key. I then tried using sops to edit the file and it complained and errored on the key/value I set. So secrets are vulnerable to bad edits, just like the rest of our deployment files which are mostly structured data of some sort.

I'm like 51/49 in favor of switching to sops. I like using git-crypt more due to how transparent it is, but the shared secret and gpg authorization mechanisms would be too unwieldy with many secrets editors. However its not clear we need to have many secrets editors because your analysis showed most of the secrets are related to cluster provisioning.

**Comment by yuvipanda** [comment]
Thanks @ryanlovett!

> I was able to create a new secret file in secrets/. So "If the user adds a new file to a secrets/ directory expecting it to be kept secret, they are in for a surprise!" is not true via your .sops.yaml glob as you indicated later. I tried using sops to create a new file outside of secrets/ but it properly errored.

Right, if you use 'sops' to create the file. If you just use vim, it'll be unencrypted. But we don't have that many files there, so it's probably not going to happen that often (famous last words!).



> However its not clear we need to have many secrets editors because your analysis showed most of the secrets are related to cluster provisioning.

Right! However, I think in the next year or two there's gonna be more people, and the secret key only has to leak once...

**Comment by betatim** [comment]
Yuvi have you used vault for secrets management, in particular together with travis? I am evaluating/working on doing this for work, somewhere along the line.

Another option i've been pondering is using a private git repository in keybase for storing the secrets. Then checking that out as part of the CI run. Keybase itself seems to be fairly well done in terms of usability (adding and removing people from a team automatically rotates keys, etc) and the team seems to take crypto serious. However I don't know how you'd cleanly get a copy into the CI (create a keybase account for your CI??)

**Comment by yuvipanda** [comment]
@betatim vault seems extremely heavy weight for this situation, and then you have to run and secure your vault server... Cloud provider's KMS can almost be used as a replacement for vault, except you don't have to run the server.

I hadn't considered keybase! I am generally not a fan of having two repos, however, since they go out of sync easily. Reverts now would require reverts in two places, and order starts to matter. You can work around this with git submodules, at the cost of the extra complexity using submodules brings.

**Comment by consideRatio** [comment]
@yuvipanda :heart: :tada: :+1: :1st_place_medal: :100: :star2: :cat2: Thank you for this write up, as often, I end up learning lots of relevant things from what you document! I love it!

**Comment by AlgoristHQ** [comment]
Can we expect an update on the Key Rotations? Is this in the works or is there a plan to implement this?

**Comment by yuvipanda** [comment]
@felder completed the move to sops (https://github.com/berkeley-dsep-infra/datahub/issues/2229), so we didn't have to do anything about rotation of git-secrets keys
